flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/BarrierBuffer.java

```diff
@@ -135,19 +138,20 @@ else if (next.getEvent().getClass() == CheckpointBarrier.class) {
 						processBarrier((CheckpointBarrier) next.getEvent(), next.getChannelIndex());
 					}
 				}
+				else if (next.getEvent().getClass() == CancelCheckpointMarker.class) {
+					processCancellationBarrier((CancelCheckpointMarker) next.getEvent());
+				}
 				else {
 					if (next.getEvent().getClass() == EndOfPartitionEvent.class) {
-						numClosedChannels++;
-						// no chance to complete this checkpoint
-						releaseBlocks();
+						processEndOfPartition(next.getChannelIndex());
 					}
 					return next;
 				}
 			}
 			else if (!endOfStream) {
 				// end of input stream. stream continues with the buffered data
 				endOfStream = true;
-				releaseBlocks();
+				releaseBlocksAndResetBarriers();
 				return getNextNonBlocked();
```

Comment:
Probably right, though it was not changed in this pull request but was like that for quite some time. 

I wanted to not touch the basic logic of this loop (it is has gone through some evolution to be stable and handle corner cases cleanly). We could look at that in a followup?